### PR TITLE
Colonel Mustard in the Lounge with the Lead Pipe (killed mapBlockIndex outside of validation)

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Replaces #10692 after #11041 was merged. Stops exporting mapBlockIndex in validation.h entirely, also making it const inside of validation.cpp outside of CChainState. Also makes chainActive const everywhere outside of CChainState and, together, removes all non-const CBlockIndex*s outside of validation/(mostly CChainState).

Sorry, no lyrics this time.

### CODE REVIEW

File: src/chain.cpp [hunk 1]
```diff
@@ -59,12 +59,17 @@ const CBlockIndex *CChain::FindFork(const CBlockIndex *pindex) const {
...
 {
     std::vector<CBlockIndex*>::const_iterator lower = std::lower_bound(vChain.begin(), vChain.end(), nTime,
         [](CBlockIndex* pBlock, const int64_t& time) -> bool { return pBlock->GetBlockTimeMax() < time; });
     return (lower == vChain.end() ? nullptr : *lower);
 }
+CBlockIndex* CChain::FindEarliestAtLeast(int64_t nTime) {
+    std::vector<CBlockIndex*>::const_iterator lower = std::lower_bound(vChain.begin(), vChain.end(), nTime,
+        [](CBlockIndex* pBlock, const int64_t& time) -> bool { return pBlock->GetBlockTimeMax() < time; });
+    return (lower == vChain.end() ? nullptr : *lower);
+}
```
Comment (Empact) [inline src/chain.cpp]: Compiles without this definition if you fix the skiplist_tests use to be const as well.

Comment (TheBlueMatt) [reply src/chain.cpp]: Done.

File: src/validation.cpp [hunk 2]
```diff
@@ -3849,26 +3849,30 @@ bool LoadChainTip(const CChainParams& chainparams)
         // that we always have a chainActive.Tip() when we return.
         LogPrintf("%s: Connecting genesis block...\n", __func__);
         CValidationState state;
-        if (!ActivateBestChain(state, chainparams)) {
+        if (!ActivateBestChain(state, chainparams, std::shared_ptr<const CBlock>())) {
```
Comment (promag) [inline src/validation.cpp]: In commit "Move LoadChainTip into CChainState",

Use `MakeShared()`?

File: src/validation.cpp [hunk 3]
```diff
@@ -191,6 +190,7 @@ class CChainState {
     CBlockIndex* FindMostWorkChain();
     bool ReceivedBlockTransactions(const CBlock &block, CValidationState& state, CBlockIndex *pindexNew, const CDiskBlockPos& pos, const Consensus::Params& consensusParams);
 
+    void PruneBlockIndexCandidates();
```
Comment (promag) [inline src/validation.cpp]: In commit "Move LoadChainTip into CChainState",

Nit, revert this unrelated change?

File: src/test/miner_tests.cpp [hunk 4]
```diff
@@ -39,6 +39,8 @@ class HasReason {
 
 static CFeeRate blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE);
 
+static CChain& nonConstChainActive = const_cast<CChain&>(chainActive);
```
Comment (promag) [inline src/test/miner_tests.cpp]: In commit "Add a const_cast'ed reference to chainActive in miner_tests",

Why add an unused variable? Could add just when it's needed?

Answer: Because later it's used in "scripted-diff: Use nonCosntChainActive in miner_tests" commit.

**Review by promag — COMMENTED** [review]
Needs rebase. Looks good.

File: src/validation.cpp (hunk 3 above)
Comment (TheBlueMatt) [reply src/validation.cpp]: How is this unrelated? Because of the change PruneBlockIndexCandidates (an internal helper function) is no longer called outside of CChainState, so it should be made private.

File: src/validation.cpp (hunk 2 above)
Comment (TheBlueMatt) [reply src/validation.cpp]: shared_ptr() creates a pointer-to-null, not a default-constructed CBlock.

File: src/validation.cpp (hunk 3 above)
Comment (promag) [reply src/validation.cpp]: Oh missed the visibility change.

**Comment by sipa** [comment]
utACK 4576f1de6725bab583797d081421b992872fd616

Nit: typo in commit name `nonCosntChainActive`.

File: src/validation.cpp (hunk 2 above)
Comment (jimpo) [reply src/validation.cpp]: Any particular reason to pass it explicitly?

File: src/validation.cpp [hunk 5]
```diff
@@ -177,6 +178,13 @@ class CChainState {
 
     void UnloadBlockIndex();
 
+    ~CChainState() {
+        BlockMap::iterator it1 = mapBlockIndex.begin();
```
Comment (jimpo) [inline src/validation.cpp]: nit: This can be simplified using the foreach loop in `UnloadBlockIndex`. And the explicit clear at the end is unnecessary.

```c++
for (BlockMap::value_type& entry : mapBlockIndex) delete entry.second;
```

**Review by jimpo — COMMENTED** [review]
Looks good. Needs rebase.

**Comment by jnewbery** [comment]
PR would be more searchable with a descriptive title :)

**Comment by Empact** [comment]
nit: use `?` in regex for 0-1 copies, rather than `*` (0-n copies)

**Comment by sipa** [comment]
Needs rebase due to #10244.

File: src/validation.cpp (hunk 2 above)
Comment (promag) [reply src/validation.cpp]: Indeed, this change could be reverted (it's the default value).

File: src/validation.cpp (hunk 5 above)
Comment (promag) [reply src/validation.cpp]: Commit "mapBlockIndex only in validation.cpp, and const outside CChainState"

> And the explicit clear at the end is unnecessary.

@jimpo why? Otherwise it would contain dangling pointers. `mapBlockIndex` isn't a `CChainState` member.

File: src/validation.cpp [hunk 6]
```diff
@@ -166,18 +169,24 @@ class CChainState {
...
     bool RewindBlockIndex(const CChainParams& params);
     bool LoadGenesisBlock(const CChainParams& chainparams);
-
-    void PruneBlockIndexCandidates();
+    bool LoadChainTip(const CChainParams& chainparams);
 
     void UnloadBlockIndex();
 
+    ~CChainState() {
+        BlockMap::iterator it1 = mapBlockIndex.begin();
```
Comment (promag) [inline src/validation.cpp]: Commit "mapBlockIndex only in validation.cpp, and const outside CChainState"

Should `AssertLockHeld(cs_main)`?

File: src/qt/clientmodel.cpp [hunk 7]
```diff
@@ -137,7 +137,7 @@ size_t ClientModel::getMempoolDynamicUsage() const
 
 double ClientModel::getVerificationProgress(const CBlockIndex* tipIn) const
 {
-    const CBlockIndex* tip = const_cast<const CBlockIndex* >(tipIn);
+    const CBlockIndex* tip = tipIn;
```
Comment (promag) [inline src/qt/clientmodel.cpp]: Commit "Remove now-useless const_cast in qt/clientmodel"

Could remove this line and rename argument to `tip`.

**Review by promag — COMMENTED** [review]
Typo in commit "scripted-diff: Use nonCosntChainActive in miner_tests"
